### PR TITLE
Hides cell toolbar when cell is collapsed

### LINKED ISSUE #13254: Cell toolbar shows up on collapsed cells even when shrunk
## Description

Cell toolbar shows up on collapsed cells even when shrunk.

## Reproduce

Screenshot from 2022-10-16 22-05-06

1. Create a second cell, enter it
2. Collapse the cell
3. See the cell toolbar stays even when it should be disabled due to space constraints (too narrow window)

## Expected behavior

Cell toolbar is either:
- not visible on collapsed cells
- not visible on collapsed cells on narrow screens to avoid visual overflow and inability to click on the expand button

## Context

- Browser and version: Chromium
- JupyterLab version: 3.x and 4.0.0a30

### PR DESCRIPTION
**Pull request by JasonWeill**
## References

Fixes #13254 (second change)

## Code changes

Modifies the code to respond to changes in cells being collapsed or expanded.

## User-facing changes

After collapsing a cell, the cell toolbar should never appear beside the placeholder.

after collapsing

After expanding a collapsed cell, the cell toolbar should appear.

after expanding

Known issue: cell toolbar may overlay content (see #13857). This change partially fixes the overlap checking functionality.

## Backwards-incompatible changes

None.

### CODE REVIEW

File: packages/cell-toolbar/src/celltoolbartracker.ts [hunk 1]
```diff
@@ -322,13 +365,21 @@ export class CellToolbarTracker implements IDisposable {
...
+    let lineRight = codeMirrorLines[0].getBoundingClientRect().left;
+    const oneChar = document.createElement('span');
+    oneChar.textContent = 'a';
+    codeMirrorLines[0].appendChild(oneChar);
+    const charWidth = oneChar.getBoundingClientRect().width;
+    codeMirrorLines[0].removeChild(oneChar);
+    // Record only the text content's length for each child node
+    codeMirrorLines[0].childNodes.forEach((childNode: ChildNode) => {
+      lineRight += (childNode.textContent?.length || 0) * charWidth;
+    });
```
Comment (krassowski) [inline packages/cell-toolbar/src/celltoolbartracker.ts:382]: Forced style calculation alert. We can do this without modifying DOM (and get more accurate results at the same time):

```suggestion
    const range = document.createRange();
    range.selectNodeContents(codeMirrorLines[0].childNodes);
    const sizes = range.getClientRects();
    lineRight += [...sizes].reduce((prev, rect) => prev + rect.width, 0);
```

**Comment by krassowski** [comment]
I tagged this for 4.0 because of changes tailored towards CodeMirror 6. I _think_ that the part reflected in title of the PR (cell collapsing fix) could still be backported to 3.6.

File: packages/cell-toolbar/src/celltoolbartracker.ts (hunk 1 above)
Comment (JasonWeill) [reply packages/cell-toolbar/src/celltoolbartracker.ts]: Thank you! The proposed line 375 gave me a compilation error for attempting to provide a node list where a single node was expected. I pushed a new change with `setStart` and `setEnd` calls instead.

Comment (krassowski) [reply packages/cell-toolbar/src/celltoolbartracker.ts]: Sorry it should be `range.selectNodeContents(codeMirrorLines[0]);` without `.childNodes`.

File: packages/cell-toolbar/src/celltoolbartracker.ts [hunk 2]
```diff
@@ -322,13 +365,22 @@ export class CellToolbarTracker implements IDisposable {
...
+
+    let lineRight = codeMirrorLines[0].getBoundingClientRect().left;
+    const range = document.createRange();
+    range.setStart(codeMirrorLines[0].childNodes[0], 0);
+    const nodeLength = codeMirrorLines[0].childNodes.length;
+    const lastNode = codeMirrorLines[0].childNodes[nodeLength - 1];
+    range.setEnd(
+      lastNode,
+      Math.max((lastNode?.textContent?.length || 0) - 1, 0)
+    );
```
Comment (krassowski) [inline packages/cell-toolbar/src/celltoolbartracker.ts:381]: Can we try?

```suggestion
    range.selectNodeContents(codeMirrorLines[0]);
```

File: packages/cell-toolbar/src/celltoolbartracker.ts [hunk 3]
```diff
@@ -322,13 +365,22 @@ export class CellToolbarTracker implements IDisposable {
...
+    const sizes = range.getClientRects();
+    lineRight += [...sizes].reduce((prev, rect) => prev + rect.width, 0);
```
Comment (krassowski) [inline packages/cell-toolbar/src/celltoolbartracker.ts:383]: Am I correct thinking that `lineRight == sizes[sizes.length - 1].right`?

Could we possibly simplify this further by just taking the last child in the line and taking its `right`, or would that fail?

I am suggesting something like:

```
lineRight = codeMirrorLines[0].lastChild.getClientRects().right;
```

Comment (JasonWeill) [reply packages/cell-toolbar/src/celltoolbartracker.ts:383]: `getBoundingClientRect` exists on an `Element`, but not on a text `Node`, so `ChildNode` doesn't include it as a callable method.

File: packages/cell-toolbar/src/celltoolbartracker.ts (hunk 2 above)
Comment (krassowski) [reply packages/cell-toolbar/src/celltoolbartracker.ts:381]: Since the other suggestion was a miss, what do you think about this one?

Comment (JasonWeill) [reply packages/cell-toolbar/src/celltoolbartracker.ts]: This works! Thanks @krassowski

**Comment by krassowski** [comment]
bot please update snapshots

**Comment by jtpio** [comment]
CI kick

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Comment by fcollonval** [comment]
bot please update galata snapshots

**Review by fcollonval — COMMENTED** [review]
@JasonWeill lost of toolbars disappear in the UI test snapshots. I could not see that when testing locally. So I guess it has to do with some asynchronicity of the code. What do you think?

**Comment by JasonWeill** [comment]
@fcollonval Thanks for updating the snapshots! Looking at them:

1. ✅ galata/test/jupyterlab/notebook-edit.test.ts-snapshots/delete-cell-jupyterlab-linux.png — this change is correct; the first line overlaps with the cell toolbar, so the toolbar should be invisible.
2. ✅ galata/test/jupyterlab/notebook-markdown.test.ts-snapshots/do-not-highlight-not-latex-jupyterlab-linux.png — this change is correct; the first line overlaps with the cell toolbar, so the toolbar should be invisible.
3. ❌ galata/test/jupyterlab/notebook-markdown.test.ts-snapshots/do-not-highlight-standalone-dollar-jupyterlab-linux.png — the first line does not overlap the toolbar, so the toolbar should be visible.
4. ❌ galata/test/jupyterlab/notebook-replace.test.ts-snapshots/replace-in-markdown-rendered-cell-jupyterlab-linux.png — the first line does not overlap the toolbar, so the toolbar should be visible.
5. ❌ galata/test/jupyterlab/notebook-run-vega.test.ts-snapshots/run-cells-dark-vega-jupyterlab-linux.png — the first line does not overlap the toolbar, so the toolbar should be visible.
6. ❌ galata/test/jupyterlab/notebook-run-vega.test.ts-snapshots/run-cells-vega-jupyterlab-linux.png — the first line does not overlap the toolbar, so the toolbar should be visible.

**Comment by JasonWeill** [comment]
I can't reproduce the test failures in the pipeline using Firefox on macOS. In my screen shots below, I see the cell toolbar as expected.

galata/test/jupyterlab/notebook-markdown.test.ts-snapshots/do-not-highlight-standalone-dollar-jupyterlab-linux.png

Image

﻿﻿﻿﻿---

galata/test/jupyterlab/notebook-replace.test.ts-snapshots/replace-in-markdown-rendered-cell-jupyterlab-linux.png

Image (1)

---

galata/test/jupyterlab/notebook-run-vega.test.ts-snapshots/run-cells-dark-vega-jupyterlab-linux.png

Image (2)

---

galata/test/jupyterlab/notebook-run-vega.test.ts-snapshots/run-cells-vega-jupyterlab-linux.png

Image (3)

@fcollonval I don't have regular access to a Linux machine; how can I debug these screenshot test failures?

**Comment by fcollonval** [comment]
I tested the following test locally:

```sh
PWDEBUG=1 jlpm test -g 'Do not enter math mode for standalone dollar'
```

It should have a toolbar but it does not. And this is not a question of asynchronicity; waiting never bring the toolbar. But activating another cell and coming back to the cell does.

Introspecting the DOM, the cell toolbar is on the expected cell but it has `display: none` style.
We may need to wait for some sort of readiness in `_codeOverlapsToolbar` before checking to display the toolbar or not.

**Comment by fcollonval** [comment]
Did you reproduce the test manually on Firefox or just changed the browser target of playwright to be Firefox?

**Comment by fcollonval** [comment]
Switching to `firefox` in playwright configuration locally, indeed the toolbar is displayed.

File: packages/cell-toolbar/src/celltoolbartracker.ts [hunk 4]
```diff
@@ -322,13 +365,17 @@ export class CellToolbarTracker implements IDisposable {
...
     }
-    const lineRight = codeMirrorLines[0].children[0] // First span under first pre
-      .getBoundingClientRect().right;
+
+    let lineRight = codeMirrorLines[0].getBoundingClientRect().left;
+    const range = document.createRange();
+    range.setStart(codeMirrorLines[0].childNodes[0], 0);
+    range.selectNodeContents(codeMirrorLines[0]);
+    const sizes = range.getClientRects();
+    lineRight += [...sizes].reduce((prev, rect) => prev + rect.width, 0);
```
Comment (krassowski) [inline packages/cell-toolbar/src/celltoolbartracker.ts:378]: Let's try this:

```suggestion
    range.selectNodeContents(codeMirrorLines[0]);
    lineRight += range.getBoundingClientRect().width;
```

After debugging the failures it seems that Chrome double-counts some nodes in the range when using `getClientRects` approach. The proposed approach works when testing in-browser.

**Comment by JasonWeill** [comment]
@krassowski Thanks for making this new commit! I see three test failures in UI tests, although for two of them, the new "actual" is actually correct. For the third, "Delete cell" (`test/jupyterlab/notebook-edit.test.ts:92`), the cell toolbar appears overlapping line 1 of the Markdown source, whereas it shouldn't appear in this situation.

**Comment by krassowski** [comment]
> For the third, "Delete cell" (test/jupyterlab/notebook-edit.test.ts:92), the cell toolbar appears overlapping line 1 of the Markdown source, whereas it shouldn't appear in this situation.

From a quick look this is not a problem with `lineRight` anymore but an issue with `toolbarLeft`. When cell gets focused the toolbar widget is not yet correctly positioned because the buttons are not actually rendered. This only happens when activating a cell, not when resizing an active cell.

toolbar-issue

It is related to #13683 but distinct.

1) A naive fix would be to add min-width to the buttons on the toolbar:
    ```css
    .jp-cell-toolbar > .jp-Toolbar-item {
      min-width: 30px
    }
    ```
    This is of course a bad solution long-term because we would get overlap in presence of longer toolbar items if an extension adds a longer item.
2) A better fix would be to wait until all toolbar buttons are rendered. This could be implemented by waiting for a specific class name to appear in the toolbar node (to avoid depending on React lifecycle).
3) The ideal fix would be to prevent re-rendering toolbar on changing cells altogether.


Looking at the code, the toolbar is re-rendered on every change of active cell:

https://github.com/jupyterlab/jupyterlab/blob/02113198e1cf453b34c4248ccfc3143c3f0f5e5d/packages/cell-toolbar/src/celltoolbartracker.ts#L107-L125

It appears that the code already tries to wait for the items to be rendered (2), but these are clearly not ready by the time we reach `this._updateCellForToolbarOverlap` in:

https://github.com/jupyterlab/jupyterlab/blob/02113198e1cf453b34c4248ccfc3143c3f0f5e5d/packages/cell-toolbar/src/celltoolbartracker.ts#L151-L188

From a quick check the render promises are added to the `promises` list correctly but they are all already fulfilled.
It seems that for this logic to work we would need to call `update` to regenerate the promises (which works locally, I will push a commit).

**Comment by krassowski** [comment]
Let's see how it works: bot please update galata snapshots. @JasonWeill is there a reason for re-rendering the toolbar when switching cells? Could we avoid it altogether (solution 3 above)? If there is a no strong counterpoint, should we open a new issue to track this as a performance enhancment?

**Comment by JasonWeill** [comment]
> Let's see how it works: bot please update galata snapshots. @JasonWeill is there a reason for re-rendering the toolbar when switching cells? Could we avoid it altogether (solution 3 above)? If there is a no strong counterpoint, should we open a new issue to track this as a performance enhancment?

@krassowski If we use the same toolbar every time, would we lose the ability to disable the move up / move down buttons, as suggested in #13502?

**Comment by fcollonval** [comment]
> @krassowski If we use the same toolbar every time, would we lose the ability to disable the move up / move down buttons, as suggested in #13502?

We will not as the buttons themself will get rerendered if the notify command changed is emitted:

https://github.com/jupyterlab/jupyterlab/blob/c6180cefc57448a92a4761b66069cdd6c6f6637c/packages/ui-components/src/components/toolbar.tsx#L841-L849

**Comment by fcollonval** [comment]
Kicking the CI

**Review by krassowski — APPROVED** [review]
Thank you @JasonWeill and thanks for the patience while we were iterating to find a clean and performant solution!

**Comment by krassowski** [comment]
Let's continue discussion on refactoring the cell-toolbar rendering in https://github.com/jupyterlab/jupyterlab/issues/14139#issue-1611970404.

**Comment by JasonWeill** [comment]
Thank you @krassowski @fcollonval for your help!

**Pull request merged**